sakai-evaluation-impl/src/java/org/sakaiproject/evaluation/logic/externals/ExternalHierarchyLogicImpl.java

```diff
@@ -267,56 +497,152 @@ public void setEvalGroupsForNode(String nodeId, Set<String> evalGroupIds) {
         if (nodeIds.length > 0) {
             if (evalHierarchyProvider != null) {
                 m = evalHierarchyProvider.getEvalGroupsForNodes(nodeIds);
-            } else {
-                List<EvalGroupNodes> l = getEvalGroupNodesByNodeId(nodeIds);
-                for (EvalGroupNodes egn : l) {
-                    Set<String> s = new HashSet<String>();
-                    String[] evalGroups = egn.getEvalGroups();
-                    for (int i = 0; i < evalGroups.length; i++) {
-                        s.add(evalGroups[i]);
-                    }
-                    m.put(egn.getNodeId(), s);
-                }
+            }
+
+            // Support for resolving eval groups based on hierarchy node rules.
+            // Hierarchy rules should be obeyed regardless of if an external provider is present or not (supplemental)
+            for( String nodeID : nodeIds )
+            {
+                m.put( nodeID, getEvalGroupsForNodeSectionAware( nodeID ) );
             }
         }
         return m;
     }
 
+    /**
+     * Utility method to extract common (section aware) behaviour.
+     * 
+     * @param nodeID
+     * @return 
+     */
+    private Set<String> getEvalGroupsForNodeSectionAware( String nodeID )
+    {
+        Set<String> groups = new HashSet<>();
+        try
+        {
+            List<HierarchyNodeRule> rules = this.getRulesByNodeID( Long.parseLong( nodeID ) );
+            for( HierarchyNodeRule rule : rules )
+            {
+                if( EvalConstants.HIERARCHY_RULE_SECTION.equals( rule.getOption() ) )
+                {
+                    groups.addAll( dao.getAllSiteIDsMatchingSectionTitle( rule.getRule() ) );
+                }
+                else
+                {
+                    groups.addAll( dao.getAllSiteIDsMatchingSiteTitle( rule.getRule() ) );
+                }
+            }
+        }
+        catch( NumberFormatException ex ) { log.warn( ex ); }
+
+        return groups;
+    }
+
     public Map<String, Integer> countEvalGroupsForNodes(String[] nodeIds) {
         Map<String, Integer> m = new HashMap<String, Integer>();
```

Comment:
Declare here but move allocation into the else block as it is only needed in that scenario